user/syscalls.S

```diff
@@ -0,0 +1,67 @@
+#include <mips/asm.h>
+#include <errno.h>
+
+    .set noreorder
+    .set noat
+
+/* This implementation uses newlib-specific errno processing. */
+#define DO_SYSCALL(num)  \
```

Comment:
I believe this works well. I'm not sure if it's more clear, especially with how `bltz $v0, syscall_failed;` does not set `$ra` so that `syscall_failed` returns from 2 nested calls at once... but that's just my opinion.

I'm pushing a commit which applies this change, but can you please finish the work on this branch? It doesn't seem I'm doing anything useful here at this point.